### DISCUSSION TITLE
ワードミュートおよびブックマークのバックエンド同期について

### DISCUSSION THREAD

**Discussion by henoya** [baisony/ucho-ten#348] [General]
現在のバックエンド同期の要望のある項目についての Discussion です。

# 現在ある、バックエンド同期したい項目

- ワードミュート
- ブックマーク

**Comment by henoya** [comment]
## 1つのミュートワードの文字数制限

### バックエンド側での文字数制限に関する制限の考え方

無制限な長さの文字列を許容すると、もちろん保存できるデータ容量の問題が発生します。
ただし、選択する保存先によって異なりますが、実際には例えば32kbとか、実用上は問題ないサイズだったりもします。
32kb = 32768バイト、UTF-8 通常の日本語文字3バイトだと 10922 文字、絵文字だと4バイトなどで 8192 文字、複合文字を考えても 6バイトで 5461文字など。

事実上、ミュートワードとして考えられるのが、ポストの本文全文対象としても 300文字なので、バックエンドとしての制限は、ほとんど無視できるかと。

### ミュートワードの文字数制限として考慮すべき要素

- ミュートワードを入力するUI上の、入力可能文字数
- ミュートワードを確認表示するときの、画面表示の時に見苦しくならない文字数、もしくは表示方法

**Reply by baisony** [reply]
### ミュートワードの上限について
文字数20
設定可能数30

**Reply by baisony** [reply]
ブックマークの上限について
同期可能数100

**Comment by henoya** [comment]
## ミュートワードの処理上の仕様について

### ポストテキスト中のワードとミュートワードが一致する条件の仕様

- ポストテキスト中の文字列に完全一致でいいのか
- 英単語などの大文字小文字対応をどうするか

**Reply by so-asano** [reply]
> 英単語などの大文字小文字対応をどうするか

こちらは常識的には case-insensitive かなとはおもいます。

> ポストテキスト中の文字列に完全一致でいいのか

これについては処理の量などを考えたときに、あまりそれ以上をやりようがない気はします。
開発サイドにもリソースが足りてるわけでもないですし。

@baisony さんはどういう意見ですかね？

**Reply by henoya** [reply]
比較時にToLowercaseするとかですかね。
保存自体は入力されたままの方がいいのかな？

**Reply by so-asano** [reply]
> 比較時にToLowercaseするとかですかね。

まあそういう実装になりそうですよね（というかほかの実装を知らない…）。

> 保存自体は入力されたままの方がいいのかな？

これはそうなのかな、っておもいました。
ユーザとしては自分の入力したミュートワードが小文字になっていたり大文字になっていたりしたら違和感ありそうですよね。

**Reply by baisony** [reply]
@henoya @so-asano 

> 保存自体は入力されたままの方がいいのかな？

そうですね。

> 英単語などの大文字小文字対応をどうするか

(case-insensitiveを知らなかったんですが) case-insensitiveで実装します。

ありがとうございます。

**Comment by henoya** [comment]
## バックエンド同期に使用を検討している クラウドDBサービス

- TiDB Cloud Serverless
  - https://www.pingcap.com/tidb-serverless/
  - https://www.pingcap.com/pricing/
  - 無料枠
    - 容量： 5GB + 50M(5000万アクセス相当) x 5クラスタ
    - 無料枠オーバーしても、いきなり停止ではなく、低速モードで稼働はできるもよう
  - 無料状態でもチーム参加可能
- Vercel からのアクセスのパッケージが用意されている。
  - `@tidbcloud/serverless`
  - https://docs.pingcap.com/tidbcloud/serverless-driver
  - Vercel のような実行時のIPアドレスを固定できない環境でも使用できる

**Reply by so-asano** [reply]
ご説明をよませていただいたかぎりでは、ぼくにはこのサービスの選択について異論はないです。 @baisony さんはいかがですかね。

**Reply by baisony** [reply]
私も TiDB Cloud Serverless を用いた実装で異論はなく、このまま進めていきます。

**Reply by henoya** [reply]
では、TiDB Serverless で構成してみます。
@baisony @so-asano 
お2人を今のチームに招待して、ばいそにさんをプロジェクトオーナーにしたいと思います。
まだアカウント登録しただけでカードもなんにも登録していないので、そこらへんは自由っぽいです。
招待を出しますので、メールアドレスを教えてください。

**Reply by so-asano** [reply]
@henoya 

了解です。Discord のメッセージでメールアドレスをお送りしますので、よろしくお願いします。

**Comment by henoya** [comment]
## TiDB で作成してみた ワードミュートとブックマークのテーブル

```sql
/* ワードミュートテーブルを作成 */
create table `ucho_ten`.`wordmutes` (
  `id` bigint(20) unsigned  NOT NULL AUTO_INCREMENT COMMENT 'id',
  `did` VARCHAR(64) not null default "" COMMENT 'did',
  `muteword` text CHARACTER SET utf8mb4 COLLATE utf8mb4_bin COMMENT 'ワードミュート情報 json文字列',
  `created_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP COMMENT 'レコード作成日時', 
  `updated_at` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP, 
  `deleted_at` DATETIME comment 'レコード削除日時', 
  PRIMARY KEY (`id`),
  UNIQUE KEY `uidx_wordmute_did` (`did`)
) AUTO_INCREMENT = 1 ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='ワードミュートテーブル';

/* ブックマークテーブルを作成 */
create table `ucho_ten`.`bookmarks` (
  `id` bigint(20) unsigned  NOT NULL AUTO_INCREMENT COMMENT 'id',
  `did` VARCHAR(64) not null default "" COMMENT 'did',
  `bookmark` text CHARACTER SET utf8mb4 COLLATE utf8mb4_bin comment 'ブックマーク情報 json文字列',
  `created_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP comment 'レコード作成日時', 
  `updated_at` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP, 
  `deleted_at` DATETIME comment 'レコード削除日時', 
  PRIMARY KEY (`id`),
  UNIQUE KEY `uidx_bookmark_did` (`did`)
) AUTO_INCREMENT = 1 ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='ブックマークテーブル';
```

基本的には、`did` をユニークキーとして、必要な情報を全てjson文字列にして、一つのカラムに入れるという方式です。
DBとしての検索性は、`did`カラムをキーとする以外には持たせていません。

個別のワードなどの削除処理も、全てプログラム側の処理を前提としています。

## 疑問点

@baisony 
- ワードミュートとブックマークを、あえて分けて管理する意味はありますでしょうか？
- 逆に、ワードミュートの項目数は、利用者側でもある程度の設定限度は出てくるでしょうが、ブックマークは、増える一方になると思うので、保存はともかく、プログラム側での処理の負荷が増加していく懸念があります。
- 一定以上アクセスなかったユーザーのデータを削除するということが、話で上がっていましたが、そのためには何らかのかたちで、`最終アクセス日時` を記録しておかないと、判定に基準にならないと思われます。
  - `last_loggedin` テーブルのような情報は必要ではないでしょうか。

**Reply by so-asano** [reply]
> ブックマークは、増える一方になると思うので、保存はともかく、プログラム側での処理の負荷が増加していく懸念があります。

これ、たしかに大きな問題なんですよね。シチュエーションが違うとはいえ、同様の問題にぶつかったことがあります（先にかいておきますが、ぼくはいいかんじには解決できていないので良い解を知りません）。
どの程度のブックマーク数を想定しているのかで対応も変わってくるとおもうんですが、 @baisony さんその点いかがです？
これは、運用の期間も関係してくる話だとおもいます。期間をとおしてブックマークの数は増えていく一方だと想定されるので。

**Reply by baisony** [reply]
@so-asano 
ブックマーク上限数に関してはおっしゃる通りで、ブックマークの数に上限を設けようと思っています。
具体的には50-100の間を想定しています。

**Reply by so-asano** [reply]
@baisony 
これどういう仕様にするんですか？
たとえば、古いブックマークは自動削除、あるいは、既存のブックマークを削除しないとあらたにブックマークできない、など。

**Reply by baisony** [reply]
@so-asano 
> 既存のブックマークを削除しないとあらたにブックマークできない

この仕様を考えています

**Reply by so-asano** [reply]
@baisony 
その場合、次のようなことが必要になりますね:
- 上限に達しているときはブックマークを不可とする。アラートを出したりしたほうがいいかもしれない。シンプルなのはボタンを表示しない、とか。

あと、次のようなことも考えられます:
- ブックマーク画面でブックマークが上限数に達していれば、その旨を表示する。
- ブックマーク画面に n / m みたいな数の表示をおこなう。

**Reply by baisony** [reply]
@henoya last_loggedinについて、それも一つのテーブルかつ、その中のカラム？として格納できますでしょうか？
でしたらこちらの仕様でお願いしたいです。

**Comment by baisony** [comment]
@henoya さん　ありがとうございます。

## ワードミュートとブックマークを、あえて分けて管理する意味はありますでしょうか？
自分はほぼSQLを触ったことがない、かつ、知識もほとんどないという前提でお話しします。
ほぼ知識がない自分からすると、テーブルを分けるメリット、デメリットをあまり理解しておらず、自分の理解の範疇で整理させてください。

### デメリット
例えば今の感じで、分けることを仕様とするデメリットは「通信回数の増加」が挙げられるのでしょうか。
月間5000万回の同期まで無料とのことですが、1回の同期につき2つDBにアクセスすること、すなわち実質的に2500万回の同期が無料枠の範囲になるという理解であっていますでしょうか。

### メリット
同期項目を増やす際、やりやすくなる

といった感じでしょうか。

### 疑問点
仮に、上記のデメリットの理解があっているならば、一つのテーブルで管理する場合におけるフィールド追加は簡単にできるのでしょうか。

## last_loggedin テーブルのような情報は必要ではないでしょうか。
提案いただきありがとうございます。そう思います。

**Reply by henoya** [reply]
処理行数としては倍にはなるので、1回で済ませた方が節約にはなると思います。

別途ブックマークの件数については、バックエンド側よりもプログラム側でどれだけの数が扱えるかは検討が必要だと思います。

**Reply by baisony** [reply]
なるほど。ではこちらの理解であっている、ということですね。ありがとうございます。

ブックマークに関してもミュートワード同様、登録可能件数に上限を設けた方が良いかもしれませんね。50件とか100件とか。

**Reply by so-asano** [reply]
> 一つのテーブルで管理する場合におけるフィールド追加は簡単にできるのでしょうか

フィールド追加のマイグレーションはたいていの場合対して面倒じゃないだろうけど、そもそもマイグレーション自体しなくていいように json をまとめてひとつのフィールドに保存してしまえばいいんじゃないかとぼくはおもってます。

**Reply by henoya** [reply]
一応用語説明
マイグレーション:
狭い意味では、DBに対してテーブルの追加・削除や、テーブルのカラム追加や設定変更、削除をおこなうことを指します。
データ構造が変わるので、一般的にはプログラム側の変更も必要になります。
また、データ構造が変わることによって、これまでのレコードに存在しなかったカラムが追加されたり、逆にデータがあったカラムのデータをどうするのか(捨てるのか、ほかになんとか振り分けるのか)などの問題も起こります。
データの問題抜きにしても、プログラムとセットで変更しないと誤動作するので、テーブルのデータ構造自体をプログラムと一緒にGit管理することも一般におこなわれます。

**Reply by baisony** [reply]
@henoya 
テーブルはまとめていただく方針でお願いします。
soさんより「テーブルをまとめる場合、フィールドもまとめるか」
```typescript
{
  "bookmark": ~,
  "muteword": ~
}
```
という案をいただいたのですが、こちらはいかがでしょうか。
問題ないようであれば、これらの方針で進めていただきたいです。

**Comment by baisony** [comment]
ここで新たな疑問点が浮かんできました。

有頂天はアカウント切り替え機能を提供しています。
異なるDIDで同じ内容のブックマーク / ミュートワードがDB内に格納され、逼迫してしまうのではないか、ということです。
localStorageのbookmark / muteWordsというキー名に配列として入れており、DID毎に管理しているわけではありません。

考えられるケースとして、
- AというDIDでブックマーク / ミュートワードを追加
- DIDがAというデータでDBと同期が発生
- BというDIDへアカウント切り替え
- Bでミュートワードやブックマークを追加
- その際に Aで同期されたモノ + Bで新しく追加したモノの同期処理が発生

このように、切り替えアカウントを増やせば増やすほどDBに入るデータが多くなるのではないか、ということです。

そこで、この問題が起こりうるかどうか、問題が起きるのであればどのような解決方法があるか、を教えていただければ嬉しいです。

**Reply by henoya** [reply]
そもそも、アカウント切り替え(didの変更)によるワードミュートリスト、ブックマークリストの共有(アカウントを切り替えても同じワードミュートリスト、ブックマークリストを共有できる)のだとすると、同じリストを積極的に重複記録するか、別のアカウントグループIDのようなもの、もしくはどれかひとつのdidで複数のdidを代表する仕組みなどが必要そうですが、その場合、そのアカウントグループ設定自体も同期対象としないと、破綻しますね。
別のブラウザでも同じアカウントグループを使ってもらわないと、齟齬が出てきます。

**Reply by baisony** [reply]
一番スマートなやり方としては、ミュートワード、ブックマークをlocalStorageにてDID毎に管理するように仕様を変更することでしょうね、、、。

クライアントサイドでの管理方法を変更すること以外コストはかからないでしょうから、この方法に変更しようと思います。

ありがとうございます。
